### PR TITLE
Gem implementation

### PR DESCRIPTION
**Pull request by akilmadan**

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -1 +1,100 @@
-# TODO: Create Readme
+# HumanizedId
```
Comment (msmirnoff) [inline Readme.md:1]: This is a very well written Readme.
Fantastic stuff.

File: lib/humanized_id.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+module HumanizedId
+  DEFAULT_GENERATION_LENGTH = 24
+  CHARACTERSET =
+    (('0'..'9').to_a + ('A'..'Z').to_a - %w(0 1 5 8 A B E I L O S U)).freeze
```
Comment (msmirnoff) [inline lib/humanized_id.rb]: For users of the gem, we may wish to note WHY this is the charset.
The choice of charset is entirely arbitrary.
The question is, what is 'easy to read' and what is 'vulgar'?
E.G. In this case, `0 1 5 8 O I L S B` are removed so that the result is easy to read, as the letters and numbers in that list are hard to tell apart.
The rest, the removed vowels, are dropped because English profanity is made of words with `U`, and general impolite words typically contain `A`, and occasionally `E`.

File: lib/humanized_id/humanizer.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+require 'facets'
+
+module HumanizedId
+  class Humanizer
+    def initialize(id:, length: nil, prefix: '')
+      @id               = id
+      @length           = length || id.to_s.length
+      @prefix           = prefix
```
Comment (msmirnoff) [inline lib/humanized_id/humanizer.rb]: If someone passes `prefix: nil` this won't actually set it to the empty string.
However, as it's only used in the string interpolation it works fine.

File: lib/humanized_id/humanizer.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    def generate_humanized_id
+      new_id = translate_id_to_charset
+      new_id = resize(id: new_id)
+      "#{@prefix}#{new_id}"
+    end
+
+    private
+
+    def translate_id_to_charset(id: @id, target_charset: @target_charset)
```
Comment (msmirnoff) [inline lib/humanized_id/humanizer.rb:18]: Neat. So effectively this passes instance variables by value into local variables. But why is this necessary, why not just use the instance variables? `to_s` is not destructive, neither is picking a new char.

File: lib/humanized_id.rb (hunk 2 above)
Comment (mnoack) [inline lib/humanized_id.rb]: yeah, you could almost have
NUMBERS_LETTERS_SIMILAR = [0 O 1 I L 5 S 8 B]
VOWELS = [A E I O U]
Then CHARACTERSET = 0-9 + A-Z - VOWELS - NUMBERS_LETTERS_SIMILAR

File: Readme.md (hunk 1 above)
Comment (akilmadan) [inline Readme.md:1]: Cheers @msmirnoff !

File: lib/humanized_id.rb (hunk 2 above)
Comment (akilmadan) [inline lib/humanized_id.rb]: Good idea :)

File: lib/humanized_id/humanizer.rb (hunk 4 above)
Comment (akilmadan) [inline lib/humanized_id/humanizer.rb:18]: Agreed - the action on the variables aren't destructive. This was more of a nicety I gained by using named parameters. It allows us to use either dependency injection or default to instance variables. Not so useful for this method, but I use it in the next method pretty effectively (using the 'new_id' as a dependency injection)

File: lib/humanized_id/humanizer.rb (hunk 3 above)
Comment (akilmadan) [inline lib/humanized_id/humanizer.rb]: Heh - good point. I'll fix it anyway

File: lib/humanized_id/humanizer.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      return char if target_charset.include? char
+      target_charset[SecureRandom.random_number(target_charset.size)]
+    end
+
+    def resize(id: @id, required_length: @length)
+      if required_length > id.length
+        # Pad with random humanized characters
+        id = "#{HumanizedId.generate_random(length: (required_length - id.length))}#{id}"
+      elsif required_length < id.length
+        id = id.slice(0..(required_length - 1))
```
Comment (msmirnoff) [inline lib/humanized_id/humanizer.rb]: Why the reduction by 1?
If it needs by to be longer we pad it, ok, but if it needs to be shorter than we reduce it by one less than what it needs to be?

File: lib/humanized_id/humanizer.rb (hunk 4 above)
Comment (msmirnoff) [inline lib/humanized_id/humanizer.rb:18]: Cool :shipit:

File: lib/humanized_id/humanizer.rb (hunk 5 above)
Comment (msmirnoff) [inline lib/humanized_id/humanizer.rb]: And what happens when `required_length == id.length`?

Comment (akilmadan) [inline lib/humanized_id/humanizer.rb]: We're using indexes starting at 0. So if we want length of 4, we'll have to use indexes 0,1,2,3.
The id is untouched if required_length == id.length and is returned in its original format.

Comment (msmirnoff) [inline lib/humanized_id/humanizer.rb]: The id is returned, you are quite right.

Comment (msmirnoff) [inline lib/humanized_id/humanizer.rb]: But for indexes, I'm still missing a piece of the puzzle.

irb(main):010:0> 'abcde'.slice 0, 3

=> "abc"

File: spec/humanized_id/humanizer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    context 'testing various ids' do
+      context 'with human friendly characters' do
+        let(:params) { { id: 'CDFGHKMN' } }
+        it 'should return original string' do
+          expect(generated_id).to eq params[:id]
+        end
+      end
+
+      context 'with non human friendly characters' do
+        let(:params) { { id: '22FARTYFART' } }
```
Comment (msmirnoff) [inline spec/humanized_id/humanizer_spec.rb:18]: Fantastic.

File: lib/humanized_id/humanizer.rb (hunk 5 above)
Comment (akilmadan) [inline lib/humanized_id/humanizer.rb]: Ah yes - I see now, many ways to skin the cat. 

slice(index) → new_str or nil click to toggle source
slice(start, length) → new_str or nil
slice(range) → new_str or nil
slice(regexp) → new_str or nil
slice(regexp, capture) → new_str or nil
slice(match_str) → new_str or nil

I do like it your way.

File: spec/humanized_id/humanizer_spec.rb (hunk 6 above)
Comment (mnoack) [inline spec/humanized_id/humanizer_spec.rb:18]: Alessandro would say this is a code "smell" :p - Fine by me :)
